File: Blobcode/inject/FileSystemHandle.js
/**
 * The FileSystemHandle interface of the File System Access API is
 * an object which represents a file or directory entry. Multiple
 * handles can represent the same entry. For the most part you do
 * not work with FileSystemHandle directly but rather its child
 * interfaces FileSystemFileHandle and FileSystemDirectoryHandle.
 */
class FileSystemHandle {
    /**
     * Create a new FileSystemHandle instance.
     * @param {string} name The name of the directory.
     * @param {string} identifier The identifier for the file system entry.
     */
    constructor(name, kind, identifier) {
        this.name = name;
        this.kind = kind;
        this.identifier = identifier;
    }

    /**
     * Compares two handles to see if the associated entries (either a file or directory) match.
     * @param {FileSystemHandle} other The FileSystemHandle to match against the handle on which the method is invoked.
     * @returns {boolean} Returns a Boolean which is true is the entries match.
     */
    isSameEntry(other) {
        return this.identifier === other.identifier;
    }

    /**
     * Queries the current permission state of the current handle.
     * @param {FileSystemHandlePermissionDescriptor} descriptor An object which specifies the permission mode to query for.
     * @returns {PS} which is one of 'granted', 'denied' or 'prompt'.
     */
    queryPermission(descriptor) {
        //TODO: Implement queryPermission
        error("NOT IMPLEMENTED: QUERY PERMISSION");
        return descriptor.mode === "read" ? "granted" : "granted";
    }

    /**
     * Requests read or readwrite permissions for the file handle.
     * @param {FileSystemHandlePermissionDescriptor} descriptor An object which specifies the permission mode to query for.
     * @returns {PS} which is one of 'granted', 'denied' or 'prompt'.
     */
    requestPermission(descriptor) {
        //TODO: Implement requestPermission
        error("NOT IMPLEMENTED: REQUEST PERMISSION");
        return descriptor.mode === "read" ? "granted" : "granted";
    }
}

File: Blobcode/FileSystemHandle.swift
//
//  PickerHandler.swift
//  Codepad
//
//  Created by Maximilian Mittelhammer on 10.12.21.
//

import Foundation

//MARK: FileSystemHandle
/**
 The FileSystemHandle interface of the File System Access API is an object which represents a file or directory entry. Multiple handles can represent the same entry. For the most part you do not work with FileSystemHandle directly but rather its child interfaces FileSystemFileHandle and FileSystemDirectoryHandle.
 */
class FileSystemHandle: Encodable {
    let kind: String
    let name: String
    let url: String
    let identifier: String
    
    init(kind: String, name: String, url: String, identifier: String){
        self.kind = kind
        self.name = name
        self.url = url
        self.identifier = identifier
    }
    
    func toJSON() -> String{
        do {
            let data = try JSONEncoder().encode(self)
            return String(data: data, encoding: .utf8)!
        } catch {
            print(error)
            return "{}"
        }
    }
}

//MARK: FileSystemFileHandle
/**
 The FileSystemFileHandle interface of the File System Access API represents a handle to a file system entry. The interface is accessed through the window.showOpenFilePicker() method.
 */
class FileSystemFileHandle: FileSystemHandle {
    init(name: String, url: String, identifier: String) {
        super.init(kind: "file", name: name, url: url, identifier: identifier)
    }
}

//MARK: FileSystemDirectoryHandle
/**
 The FileSystemDirectoryHandle interface of the File System Access API provides a handle to a file system directory. The interface is accessed via the window.showDirectoryPicker() method.
 */
class FileSystemDirectoryHandle: FileSystemHandle {
    var entries: [FileSystemHandle] = []
    
    init(name: String, url: String, identifier: String) {
        super.init(kind: "directory", name: name, url: url, identifier: identifier)
    }
    
    func append(fileHandle: FileSystemHandle){
        entries.append(fileHandle)
    }
    
    private enum CodingKeys: String, CodingKey {
        case kind
        case name
        case url
        case identifier
        case entries
    }

    override func encode(to encoder: Encoder) throws {
        try super.encode(to: encoder)
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(self.kind, forKey: .kind)
        try container.encode(self.name, forKey: .name)
        try container.encode(self.url, forKey: .url)
        try container.encode(self.identifier, forKey: .identifier)
        try container.encode(self.entries, forKey: .entries)
    }
}
